webapp/src/shared/components/Game/Game.tsx

```diff
@@ -1,78 +1,157 @@
+  const navigate = useNavigate();
+  const [gameJoined, setGameJoined] = React.useState(false);
+  const sendGameCommand = React.useCallback(
+    (command: GameCommand) => {
+      if (isPlayerRef.current && gameJoined) {
+        socket.emit(GAME_COMMAND, {
+          command,
+          gameRoomId: gameId,
+        });
+      }
+    },
+    [gameId, gameJoined],
+  );
   useGameControls(sendGameCommand);
-  const { runGameFrame } = useGameEngine(updateGame);
 
-  const gameLoop = React.useCallback(() => {
-    const canvasContext = canvasRef.current?.getContext('2d');
-    const gameState = runGameFrame();
+  const joinGame = () => {
+    if (!gameJoined) {
+      socket.emit(JOIN_GAME, { gameRoomId: gameId });
+    }
+  };
 
-    setScore(gameState.score);
-    canvasContext && renderFrame(canvasContext, gameState);
-    window.requestAnimationFrame(() => gameLoop());
-  }, [runGameFrame, renderFrame]);
+  React.useEffect(() => {
+    return () => {
+      if (gameJoined) {
+        socket.emit(LEAVE_GAME, { gameRoomId: gameId });
+      }
+    };
+  }, [gameId, gameJoined]);
 
   React.useEffect(() => {
-    const canvas = canvasRef.current;
-
-    if (canvas) {
-      canvas.width = CANVAS_WIDTH;
-      canvas.height = CANVAS_HEIGHT;
-      requestFrameRef.current = window.requestAnimationFrame(() => gameLoop());
-
-      return () => {
-        if (requestFrameRef.current) {
-          cancelAnimationFrame(requestFrameRef.current);
-        }
-      };
+    function handleGameJoined(info: GameInfo) {
+      isPlayerOneRef.current = authUser?.id === info.playerOneId;
+      isPlayerRef.current =
+        authUser?.id === info.playerOneId || authUser?.id === info.playerTwoId;
+      setGameJoined(true);
     }
-  }, [gameLoop]);
 
-  return (
-    <>
-      <h1 className="game-score heading-bold">{score}</h1>
-      <canvas className="game-arena" ref={canvasRef} />
-    </>
-  );
-};
+    function handleUpdateGame(info: GameInfo) {
+      gameStateRef.current = info.gameState;
+      if (gameStateRef.current) {
+        const gameState = gameStateRef.current;
+        const canvasContext = canvasRef.current?.getContext('2d');
+
+        setScore(gameState.score);
+        setOpponentScore(gameState.scoreOpponent);
+        canvasContext &&
+          renderMultiplayerFrame(
+            canvasContext,
+            gameState,
+            isPlayerOneRef.current,
+          );
+      }
+    }
+
+    function handleGameNotFound() {
+      navigate(PLAY_URL, { replace: true });
+    }
 
-const Start = () => {
-  const { send } = useStateMachineContext();
+    function handleGameEnded() {
+      notify('Game ended');
+      navigate(PLAY_URL, { replace: true });
+    }
+
+    socket.on(GAME_JOINED, handleGameJoined);
+    socket.on(UPDATE_GAME, handleUpdateGame);
+    socket.on(GAME_NOT_FOUND, handleGameNotFound);
+    socket.on(GAME_FINISHED, handleGameEnded);
+    socket.on('exception', (wsError: WsException) => {
```

Comment:
We're subscribed to the
`exception` ws message in at least three places, two of them overlapping, in two places.
We are subscribed in the GamePairingContext, and here, and in the GamePairingContext and somewhere else. I've seen that this doesn't appear to break anything, but maybe it will bring problems? Maybe we'd like to opena  ticket?

Reply:
I've removed the listener from this component. Maybe we could remove all and leave only the one from the `GamePairingContext` since it's the outermost ws socket context?

Reply:
Yes, good idea. Don't necessarily do it here and now, though.  Should I open an issue?

Reply:
Done, #489